File: tanmayC++/containerWithMostWater.cpp
#include <iostream>
using namespace std;
int main(){
    int n;
    cout<<"enter number of pillers : ";
    cin>>n;
    int hig[n];
    cout<<"enter the heights of pillers...\n";
    for(int i=0;i<n;i++){
        cout<<i+1<<" -> ";
        cin>>hig[i];
    }
    int maxar=0;
    int curar=0;
    int width;
    int minhig;
    int a,b;
    for(int i=0,j=(n-1);i<j;){
        width=j-i;
        minhig=min(hig[i],hig[j]);
        curar=width*minhig;
        if(curar>maxar){
            a=i;
            b=j;
            maxar=curar;
        }
        hig[i]<hig[j]?i++:j--;
    }
    cout<<"Max water -> "<<maxar<<" square units"<<endl;
    cout<<"between the pillers "<<a+1<<" and "<<b+1;

}

File: tanmayJava/ArrayLists/ContainerWithMostWater.java
import java.util.*;

public class ContainerWithMostWater {
    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            // Input heights
            ArrayList<Integer> height = new ArrayList<>();
            System.out.println("Enter array elements (end with '-1'):");
            int num = sc.nextInt();
            while (num != -1) {
                height.add(num);
                num = sc.nextInt();
            }

            // Handle edge cases
            if (height.size() < 2) {
                System.out.println("Not enough poles to form a container.");
                return;
            }

            // 2 pointer approach
            int st = 0;
            int en = height.size() - 1;
            // vars to store max water poles and ammount
            int p1 = st;
            int p2 = en;
            int maxWater = 0;

            while(st < en){
                // calculate current water between the pointers
                int currWater = (Math.min(height.get(st), height.get(en)))*(en-st);
                // update max water data
                if (currWater > maxWater){
                    maxWater = currWater;
                    p1 = st;
                    p2 = en;
                }
                // move the smaller pole pointer
                if (height.get(st) < height.get(en)){
                    st++;
                }
                else{
                    en--;
                }
            }
            // Display results
            System.out.println("Pole 1: " + p1 + ", Pole 2: " + p2 + ", Max Water: " + maxWater + " units");
            // time complexity O(n)
        }
    }
}
